Overhaul arch-split in machine, ASpec and AInvs

Xaphiosis opened the pull request:
In https://github.com/seL4/l4v/pull/791 I went over AInvs for AARCH64 only. I've since expanded the approach to ASpec as well, and deployed it across all the architectures. The speed benefits still apply, and now ASpec is faster as well. I'll see if I can do a better benchmark, or we can pull it out of the regression tests.

🦆🦆🦆 I have put a lot of effort into splitting up the commits into being topical, because this is a rather large PR. They should not overlap much, so if you review a line in the commit view it's unlikely to be changed in another commit.

@mbrcknl @michaelmcinerney FYI in case you are interested in the new arch-splitting tech/approach.

- [ ] add `FIXME arch_split` to PR linter to stop people reintroducing it
- [x] remove `apply_trace` I accidentally left in (`proof/invariant-abstract/X64/ArchCSpace_AI.thy` in `ainvs: disambiguate acap_rights_update_id`)

lsf37 commented on proof/invariant-abstract/ARM/ArchVSpaceEntries_AI.thy:
```diff
@@ -9,7 +9,7 @@ imports VSpaceEntries_AI
 
 
-context Arch begin global_naming ARM (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
This FIXME can probably be dropped now -- the arch split has happened.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchKernelInit_AI.thy:
```diff
@@ -14,7 +14,7 @@ imports
 begin
 
-context Arch begin global_naming ARM (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
this FIXME can also be dropped

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchRetype_AI.thy:
```diff
@@ -769,7 +766,7 @@ sublocale retype_region_proofs_gen?: retype_region_proofs_gen
 
 
-context Arch begin global_naming ARM_HYP (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
Less clear here, but I don't think there is anything that should be fixed here, so probably also to be dropped.

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVCPU_AI.thy:
```diff
@@ -9,7 +9,7 @@ theory ArchVCPU_AI
 begin
 
-context Arch begin global_naming ARM_HYP (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
FIXME can be dropped

lsf37 commented on proof/invariant-abstract/ARM_HYP/ArchVSpaceEntries_AI.thy:
```diff
@@ -8,7 +8,7 @@ theory ArchVSpaceEntries_AI
 begin
 
-context Arch begin global_naming ARM_HYP (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
FIXME can be dropped

lsf37 commented on proof/invariant-abstract/DetSchedAux_AI.thy:
```diff
@@ -8,10 +8,9 @@ theory DetSchedAux_AI
-context begin interpretation Arch .
+(* (* FIXME arch_split: check it's global on other arches *)
 requalify_facts
-  invoke_untyped_st_tcb_at
-end
+  invoke_untyped_st_tcb_at *)
```
It looks like it is, so this block should probably be removed

lsf37 commented on proof/invariant-abstract/X64/ArchKernelInit_AI.thy:
```diff
@@ -14,7 +14,7 @@ imports
 begin
 
-context Arch begin global_naming X64 (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
FIXME can be dropped

lsf37 commented on proof/invariant-abstract/X64/ArchVSpaceEntries_AI.thy:
```diff
@@ -8,7 +8,7 @@ theory ArchVSpaceEntries_AI
 begin
 
-context Arch begin global_naming X64 (*FIXME: arch-split*)
+context Arch begin arch_global_naming (*FIXME: arch-split*)
```
FIXME can be dropped

lsf37 commented on spec/abstract/ARM_HYP/ArchVSpace_A.thy:
```diff
@@ -329,7 +329,7 @@ where
 
 definition
-  vgic_update_lr :: "obj_ref \<Rightarrow> nat \<Rightarrow> ARM_A.virq \<Rightarrow> (unit,'z::state_ext) s_monad"
+  vgic_update_lr :: "obj_ref \<Rightarrow> nat \<Rightarrow> ARM_HYP_A.virq \<Rightarrow> (unit,'z::state_ext) s_monad"
```
Not sure it's worth fixing, but it looks to me like this definition is in the Arch context, so should not need the name prefix.

Xaphiosis replied:
fixed

Xaphiosis replied:
well spotted!

lsf37 reviewed (approved):
Very nice improvement in consistency and reduction of interpretations. I only spotted a few old FIXMEs that can be removed, otherwise ready to merge from my side (when the `apply_trace` is removed, the linter can be in a separate PR if you prefer).

Xaphiosis commented:
> Very nice improvement in consistency and reduction of interpretations. I only spotted a few old FIXMEs that can be removed, otherwise ready to merge from my side (when the `apply_trace` is removed, the linter can be in a separate PR if you prefer).

Thanks for spotting those FIXMEs, I cleared them out (and some others) in a separate commit, if you want to check that.
Fixed the apply_trace as well.
The linter, yes, probably in a separate PR, I don't know how to do that yet.

corlewis reviewed (commented):
Minor, but the first commit message could mention spec being updated as well.

corlewis commented on proof/invariant-abstract/ARM/ArchCSpace_AI.thy:
```diff
@@ -604,8 +604,6 @@ lemma arch_post_cap_deletion_invs:
 
-end
-
 (* is this the right way? we need this fact globally but it's proven with
    ARM defns. *)
```
Very minor, but it looks like you've removed this comment on most architectures but left it for this one. Was that intentional?

Xaphiosis replied:
nope, thanks

corlewis commented on proof/invariant-abstract/LevityCatch_AI.thy:
```diff
@@ -19,9 +19,13 @@ arch_requalify_consts
 arch_requalify_facts
   ptrFormPAddr_addFromPPtr
   aobj_ref_arch_cap
+  arch_finalise_cap_bcorres
+  prepare_thread_delete_bcorres
 
 lemmas aobj_ref_arch_cap_simps[simp] = aobj_ref_arch_cap
 
+lemmas [wp] = arch_finalise_cap_bcorres prepare_thread_delete_bcorres
```
Out of curiosity, why were these added to `LevityCatch_AI`? Is it that there's nowhere better for them? I would have thought that we would like to minimise its content, with it mainly containing lemmas that could be moved elsewhere.

Xaphiosis replied:
Definitely a derp. Would this be acceptable instead?
```
           apply (simp add: select_ext_fa[simplified free_asid_select_def]
                            free_asid_select_def o_def returnOk_liftE[symmetric]
                       split del: if_split)
```

corlewis replied:
Just to make sure, I guess this was for the conversation just below?

Xaphiosis replied:
Oh, right, yes. Sorry, I missed this one as well!

Xaphiosis replied:
I wanted to have these in a generic theory file, but couldn't find one that was in the import hierarchy in time for these to be useful. I'll take another look while I'm fixing X64, but I don't know if I'll find anyplace better. Advice welcome.

corlewis commented on proof/refine/ARM/Arch_R.thy:
```diff
@@ -887,8 +887,7 @@ shows
            apply (simp add: select_ext_fa[simplified free_asid_select_def]
-                            free_asid_select_def o_def returnOk_liftE[symmetric]
-                 split del: if_split)
+                       free_asid_select_def o_def returnOk_liftE[symmetric] split del: if_split)
```
Not sure about this, the previous indentation looks better.

corlewis reviewed (approved):
Very nice to see this improved consistency. Thank you for the efforts in splitting up the commits, that made it a lot nicer to review.

corlewis commented:
Some good looking benchmarks for this
```
master:
  N = 10
  Average ARM ASpec: (00:01:16 real, 00:02:44 cpu, 4.10GB)
  Standard Deviation: (00:00:01 real, 00:00:03 cpu, 0.31GB)
arch-split_ainvs:
  N = 10
  Average ARM ASpec: (00:01:06 real, 00:02:34 cpu, 3.96GB)
  Standard Deviation: (00:00:01 real, 00:00:03 cpu, 0.22GB)
Percentage change: -13.2%

master:
  N = 10
  Average ARM AInvs: (00:07:44 real, 00:40:56 cpu, 12.97GB)
  Standard Deviation: (00:00:18 real, 00:01:41 cpu, 2.82GB)
arch-split_ainvs:
  N = 10
  Average ARM AInvs: (00:07:01 real, 00:37:48 cpu, 14.23GB)
  Standard Deviation: (00:00:16 real, 00:01:28 cpu, 3.30GB)
Percentage change: -9.3%
```
```
master:
  N = 10
  Average ARM_HYP ASpec: (00:01:54 real, 00:03:47 cpu, 5.75GB)
  Standard Deviation: (00:00:02 real, 00:00:05 cpu, 0.78GB)
arch-split_ainvs:
  N = 10
  Average ARM_HYP ASpec: (00:01:34 real, 00:03:22 cpu, 5.91GB)
  Standard Deviation: (00:00:02 real, 00:00:05 cpu, 0.49GB)
Percentage change: -17.6%

master:
  N = 10
  Average ARM_HYP AInvs: (00:11:08 real, 00:51:48 cpu, 14.40GB)
  Standard Deviation: (00:00:12 real, 00:01:04 cpu, 2.60GB)
arch-split_ainvs:
  N = 10
  Average ARM_HYP AInvs: (00:09:57 real, 00:46:44 cpu, 13.04GB)
  Standard Deviation: (00:00:10 real, 00:00:51 cpu, 0.87GB)
Percentage change: -10.5%
```
```
master:
  N = 10
  Average X64 ASpec: (00:01:44 real, 00:03:31 cpu, 4.21GB)
  Standard Deviation: (00:00:01 real, 00:00:02 cpu, 0.56GB)
arch-split_ainvs:
  N = 10
  Average X64 ASpec: (00:01:26 real, 00:03:09 cpu, 4.30GB)
  Standard Deviation: (00:00:01 real, 00:00:01 cpu, 0.62GB)
Percentage change: -17.7%

master:
  N = 10
  Average X64 AInvs: (00:10:12 real, 00:53:56 cpu, 14.59GB)
  Standard Deviation: (00:00:19 real, 00:01:44 cpu, 0.88GB)
arch-split_ainvs:
  N = 10
  Average X64 AInvs: (00:08:52 real, 00:49:04 cpu, 12.56GB)
  Standard Deviation: (00:00:15 real, 00:01:17 cpu, 0.49GB)
Percentage change: -13.1%
```

Xaphiosis commented on proof/refine/X64/IpcCancel_R.thy:
```diff
@@ -1269,13 +1269,13 @@ lemma (in delete_one) suspend_corres:
 lemma no_fail_switchFpuOwner[wp]:
   "no_fail \<top> (X64.switchFpuOwner thread cpu)"
-  by (simp add: X64.switchFpuOwner_def Arch.no_fail_machine_op_lift)
+  by (simp add: X64.switchFpuOwner_def X64.no_fail_machine_op_lift)
```
Well spotted! Probably someone getting overzealous with adding qualification, there's indeed no need for it.

Xaphiosis replied:
Looks like removing this qualification breaks the X64 proof. I will investigate why.

Xaphiosis replied:
OK, this isn't an issue with ASpec/AInvs, this is someone doing an Arch-dependent proof outside of the Arch context, because why not. Will revert this "fix" and leave it for proper arch-splitting of Refine.

Xaphiosis commented:
@corlewis I addressed your comments, hopefully, with the latest commits (they're named after the commit I'm supposed to squash them into)
Thank you so much for the proper benchmarking! My own sad effort showed ARM_HYP being slower in AInvs, which was a disappointment. Speaking of disappointments, do you have any guesses why ARM got comparatively poorer improvements? I didn't notice anything relevant during this pass... `context begin interpretation Arch .` is now removed completely from `spec/abstract` and `proof/invariant-abstract`, and that was the bulk of the speedup.

corlewis commented:
> Speaking of disappointments, do you have any guesses why ARM got comparatively poorer improvements? I didn't notice anything relevant during this pass... `context begin interpretation Arch .` is now removed completely from `spec/abstract` and `proof/invariant-abstract`, and that was the bulk of the speedup.

My only guess is based on the fact that ARM seems to have the smallest spec and proof. Maybe that means its arch-specific files are smaller, and that interpreting its `Arch` locale wasn't as slow as compared to the other arches?

Xaphiosis commented:
> > Speaking of disappointments, do you have any guesses why ARM got comparatively poorer improvements? I didn't notice anything relevant during this pass... `context begin interpretation Arch .` is now removed completely from `spec/abstract` and `proof/invariant-abstract`, and that was the bulk of the speedup.
> 
> My only guess is based on the fact that ARM seems to have the smallest spec and proof. Maybe that means its arch-specific files are smaller, and that interpreting its `Arch` locale wasn't as slow as compared to the other arches?

My impression was that RISCV64 is the smallest?

Xaphiosis commented:
Squashed

corlewis commented:
> > > Speaking of disappointments, do you have any guesses why ARM got comparatively poorer improvements? I didn't notice anything relevant during this pass... `context begin interpretation Arch .` is now removed completely from `spec/abstract` and `proof/invariant-abstract`, and that was the bulk of the speedup.
> > 
> > 
> > My only guess is based on the fact that ARM seems to have the smallest spec and proof. Maybe that means its arch-specific files are smaller, and that interpreting its `Arch` locale wasn't as slow as compared to the other arches?
> 
> My impression was that RISCV64 is the smallest?

Sorry, by smallest I meant quickest. I'll also test RISCV64 to see how it compares, I didn't get around to it before.

corlewis commented:
> I'll also test RISCV64 to see how it compares, I didn't get around to it before.

I've added RISCV64 results to my earlier comment. Like you thought, it's the shortest/quickest AInvs. I don't know how much to read into it, but it does also show one of the smallest improvements.

Xaphiosis commented:
> > I'll also test RISCV64 to see how it compares, I didn't get around to it before.
> 
> I've added RISCV64 results to my earlier comment. Like you thought, it's the shortest/quickest AInvs. I don't know how much to read into it, but it does also show one of the smallest improvements.

Cool, thank you! The improvement on it still beat ARM somehow, especially in ASpec. I remain puzzled.
